fun/management/commands/export_course_anonymized_demographics.py

```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+from django.contrib.auth.models import User
+from django.core.management.base import BaseCommand, CommandError
+from django.conf import settings
+from opaque_keys.edx.keys import CourseKey
+from opaque_keys import InvalidKeyError
+from optparse import make_option
+from student.models import CourseEnrollment
+from student.models import UserProfile
+from xmodule.modulestore.django import modulestore
+import hashlib
+import hmac
+import sys
+import logging
+
+logger = logging.getLogger(__name__)
+#pour ques les anonymized_username soient les meme que dans les tracking log, il faudra utiliser la même ANONYMIZATION_KEY que pour ceux-ci
+
+class Command(BaseCommand):
+    help = """Gather demographics from a course_id, anonymize users and prints the result as a JSON """
+    option_list = BaseCommand.option_list + ( make_option('--course_id', action='store', dest='COURSE_ID', type='string' ) , )
+
+    def get_courses(self):
+        store = modulestore()
+        return store.get_courses()
+
+    def handle(self, *args, **options):
+        if not all([options['COURSE_ID']]):
+            raise CommandError('--course_id mandatory')
+        try:
+            course_key = CourseKey.from_string(options['COURSE_ID'])
+            if not modulestore().get_course(course_key, depth=2):
+                raise Exception
+        except InvalidKeyError:
+            print("Course id {} could not be parsed as a CourseKey;".format(options['COURSE_ID']))
+        except :
+            print  >> sys.stderr, "\n Course {} not found.".format(options['COURSE_ID'])
+            sys.exit(-1)
+        # gathers demographic data for course students
+        demographics=UserProfile.objects.filter(user__courseenrollment__course_id=course_key)
+        print '{"demographic":['
+        for i, demographic in enumerate(demographics):
+            # anonymise usernam before printing
+            anonymized_username = hmac.new(settings.ANONYMIZATION_KEY.encode(), str(demographic.user).encode('utf-8'),hashlib.sha256).hexdigest()
+            separator=','
+            if i == len(demographics) - 1:
+                separator=""
+            # aucun champ ne peut contenir de '"' (double quote)
+            print '\t{{ "username": "{}", "gender": "{}", "year_of_birth": "{}", "level_of_education": "{}", "country": "{}" }}{}'\
+                .format(anonymized_username, demographic.gender, demographic.year_of_birth, demographic.level_of_education, demographic.country, separator)
+        print ']}'
```

Comment:
Si je comprends bien, on cherche à écrire du json dans la sortie standard. Nous n'avons pas à convertir nous-même des données, en json, le package `json` s'en occupe tout seul :

    import json
    ...
    def handle(self, *args, **options):
        if not all([options['COURSE_ID']]):
            raise CommandError('--course_id mandatory')
        try:
            course_key = CourseKey.from_string(options['COURSE_ID'])
            if not modulestore().get_course(course_key, depth=2):
                raise Exception
        except InvalidKeyError:
            print("Course id {} could not be parsed as a CourseKey;".format(options['COURSE_ID']))
        # gathers demographic data for course students
        demographics = UserProfile.objects.filter(user__courseenrollment__course_id=course_key)

        json_data = {
            "demographic": [profile_demographics(profile) for profile in demographics]
        }
        self.stdout.write(json.dumps(json_data, indent=2))
        self.stdout.write("\n")

    def profile_demographics(profile):
        anonymized_username = hmac.new(
        settings.ANONYMIZATION_KEY, profile.user.name, hashlib.sha256
    ).hexdigest()
    return {
        "username": anonymized_username,
        "gender": profile.gender,
        'year_of_birth': profile.year_of_birth,
        'level_of_education': profile.level_of_education,
        'country': profile.country,
    }
    data = {
    }